window_management/compass_control.py

```diff
@@ -0,0 +1,2680 @@
+# swapping west for east, left for right, etc.
+#
+#
+# - behavior while stretching or shrinking a rectangle which is not fully contained by the parent
+# rectangle is a bit funky...not terrible, I guess...but could be improved.
+#
+# - is this a bug, the fact that the settings below seem to be constantly changing? see 'settings register' below
+# and code.py.
+# 2021-12-08 22:59:37    IO CompassControl.refresh_settings: arg='user.code_public_function_formatter'
+# 2021-12-08 22:59:37    IO CompassControl.refresh_settings: arg=<talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>
+# 2021-12-08 22:59:37    IO CompassControl.Mover.refresh_settings: args=('user.code_public_function_formatter', <talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>)
+# 2021-12-08 22:59:37    IO refresh_settings: self.settings_map={'continuous_move_frequency_str': 'user.win_move_frequency', 'continuous_resize_frequency_str': 'user.win_resize_frequency', 'continuous_move_rate': 'user.win_continuous_move_rate', 'continuous_resize_rate': 'user.win_continuous_resize_rate', 'verbose_warnings': 'user.win_verbose_warnings'}
+# 2021-12-08 22:59:37    IO refresh_settings: args: args=('user.code_public_variable_formatter', <talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>)
+# 2021-12-08 22:59:37    IO CompassControl.refresh_settings: arg='user.code_public_variable_formatter'
+# 2021-12-08 22:59:37    IO CompassControl.refresh_settings: arg=<talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>
+# 2021-12-08 22:59:37    IO CompassControl.Mover.refresh_settings: args=('user.code_public_variable_formatter', <talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>)
+# 2021-12-08 22:59:37    IO CompassControl.Sizer.refresh_settings: args=('user.code_public_variable_formatter', <talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>)
+# 2021-12-08 22:59:37    IO refresh_settings: self.settings_map={'continuous_move_frequency_str': 'user.win_move_frequency', 'continuous_resize_frequency_str': 'user.win_resize_frequency', 'continuous_move_rate': 'user.win_continuous_move_rate', 'continuous_resize_rate': 'user.win_continuous_resize_rate', 'verbose_warnings': 'user.win_verbose_warnings'}
+# 2021-12-08 22:59:37    IO refresh_settings: args: args=('user.code_protected_function_formatter', <talon.scripting.types.SettingDecl.NoValueType object at 0x0000000005B7C160>)
+
+import logging
+import math
+import threading
+import time
+from typing import Any, Callable, Iterator, Optional
+
+from talon import Context, Module, cron, ctrl, settings, ui
+
+## talon stuff
+
+mod = Module()
+
+# # context used to enable/disable the tag for controlling whether the 'stop' command is active
+ctx = Context()
+
+# taken from https: //talon.wiki/unofficial_talon_docs/#captures
+#
+# a type for representing compass directions
+Direction = dict[str, bool]
```

Comment:
this seems like a somewhat cumbersome way to represent compass directions? not sure what a better one is since I haven't read all of the code that uses it.

Reply:
Perhaps, not sure I agree. In any case, I just lifted that from the Talon wiki, as indicated by the comment at line 90.